common/functions/src/scalars/maths/round.rs

```diff
@@ -0,0 +1,167 @@
+        let x_series = columns[0]
+            .column()
+            .to_minimal_array()?
+            .cast_with_type(&DataType::Float64)?;
+
+        let column: DataColumn = if columns.len() == 1 {
+            x_series.f64()?.apply(|x| (self.rounding_func)(x))
+        } else {
+            let d_column: &DataColumn = &columns[1].column().cast_with_type(&DataType::Int64)?;
+            match d_column {
+                DataColumn::Constant(v, _) => {
+                    if v.is_null() {
+                        // mysql> SELECT ROUND(1.015, NULL); => NULL
+                        DFFloat64Array::full_null(input_rows)
+                    } else {
+                        match v.as_i64() {
+                            Err(_) => {
+                                // mysql> SELECT ROUND(1.015, 'aaa'); => 1 => d == 0
+                                x_series.f64()?.apply(|x| (self.rounding_func)(x))
+                            }
+                            Ok(v) => match v.cmp(&0) {
+                                Ordering::Greater => {
+                                    let z = 10_f64.powi(if v > 30 { 30 } else { v as i32 });
+                                    x_series.f64()?.apply(|x| {
+                                        x.trunc() + (self.rounding_func)(x.fract() * z) / z
+                                    })
+                                }
+                                Ordering::Less => {
+                                    let z = 10_f64.powi(if v < -30 { 30 } else { -v as i32 });
+                                    x_series.f64()?.apply(|x| (self.rounding_func)(x / z) * z)
+                                }
+                                Ordering::Equal => {
+                                    x_series.f64()?.apply(|x| (self.rounding_func)(x))
+                                }
+                            },
+                        }
+                    }
+                }
+                DataColumn::Array(d_series) => {
+                    let round = |x: &f64, d: &i64| {
+                        let v = *d;
+                        match v.cmp(&0) {
+                            Ordering::Greater => Some({
+                                let z = 10_f64.powi(if v > 30 { 30 } else { v as i32 });
+                                x.trunc() + (self.rounding_func)(x.fract() * z) / z
+                            }),
+                            Ordering::Less => Some({
+                                let z = 10_f64.powi(if v < -30 { 30 } else { -v as i32 });
+                                (self.rounding_func)(x / z) * z
+                            }),
+                            Ordering::Equal => Some((self.rounding_func)(*x)),
+                        }
+                    };
+
+                    let opt_iter = x_series
+                        .f64()?
+                        .into_iter()
+                        .zip(d_series.i64()?.into_iter())
+                        .map(|(x, d)| match (x, d) {
```

Comment:
validity can be combined at last.

try `new_from_iter_validity`

Reply:
After #2944, we can use binary function.

Reply:
👍 
switched to binary